compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCheckNotNullCallChecker.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirAnonymousFunction
+import org.jetbrains.kotlin.fir.expressions.FirCallableReferenceAccess
+import org.jetbrains.kotlin.fir.expressions.FirCheckNotNullCall
+import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.canBeNull
+import org.jetbrains.kotlin.fir.types.coneTypeSafe
+
+object FirCheckNotNullCallChecker : FirBasicExpressionChecker() {
+    override fun check(expression: FirStatement, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirCheckNotNullCall) return
+
+        val argument = expression.argumentList.arguments.singleOrNull() ?: return
+        if (argument is FirAnonymousFunction && argument.isLambda) {
+            reporter.reportOn(expression.source, FirErrors.NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION, context)
+            return
+        }
+        if (argument is FirCallableReferenceAccess) {
+            reporter.reportOn(expression.source, FirErrors.NOT_NULL_ASSERTION_ON_CALLABLE_REFERENCE, context)
+            return
+        }
+        if (argument is FirResolvedQualifier) {
+            // Would be reported as other diagnostics, e.g., TYPE_PARAMETER_IS_NOT_AN_EXPRESSION, NO_COMPANION_OBJECT, etc.
+            return
+        }
+
+        // TODO: should be resolved type ref! Maybe same TODO at FirTypeRef.canBeNull in FirTypeUtils.kt
+        val type = argument.typeRef.coneTypeSafe<ConeKotlinType>() ?: return
```

Comment:
What is the difference with just checking `argument.typeRef.canBeNull` directly? IIUC the difference is that not-null check on expressions that still have implicit type ref would get `UNNECESSARY_NOT_NULL_ASSERTION`

Maybe you can instead change `FirTypeRef.canBeNull` to `...?.canBeNull != false`?

Reply:
Yes, that's true, and since it's a warning, adding that wouldn't be harmful (yet).

I just don't want to hide an issue---remaining implicit type ref in `FirCheckNotNullCall` here. This place should be simply `argument.typeRef.coneType` after resolving that case. Note that `FirTypeRef.canBeNull` is supposed to use `coneType` too, but then stepped back since there were still implicit type ref. I did some cleanup a while ago, but it seems like we still have some.

Reply:
It seems like the case I saw was resolved qualifier with no value type: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ResolveUtils.kt#L147

This part is now using `coneType` directly. Also, changed to use fully expanded type (in case of type alias).